### PR TITLE
Add support for display triggers in inapp notifications

### PR DESCRIPTION
**Pull request by madster26**
**Event Triggered InApps**

Currently users have no control over when an InApp notification shows up. With this release users can now control when an InApp gets displayed based on an event that they track within their app. This "trigger" event is defined during message creation. You can additionally filter the event based on properties that are tracked along with the event for even finer controls.

### CODE REVIEW

File: Mixpanel/InAppNotification.swift [hunk 1]
```diff
@@ -55,7 +57,23 @@ class InAppNotification {
...
-            Logger.error(message: "invalid notification extra section")
+            Logger.error(message: "invalid notification extra section \(object)")
+            return nil
+        }
+        
+        if let rawDisplayTriggers = object[PayloadKey.displayTriggers] as? [[String: Any]]? {
+            if let rawDisplayTriggers = rawDisplayTriggers {
+                for rawDisplayTrigger in rawDisplayTriggers {
+                    if let displayTrigger = DisplayTrigger(JSONObject: rawDisplayTrigger) {
+                        self.displayTriggers.append(displayTrigger)
```
Comment (zihejia) [inline Mixpanel/InAppNotification.swift]:
```suggestion
                        displayTriggers.append(displayTrigger)
```

Comment (zihejia) [reply Mixpanel/InAppNotification.swift]: Code style: in swift, preferably not adding self. unless necessary

File: Mixpanel/InAppNotification.swift [hunk 2]
```diff
@@ -110,6 +145,11 @@ class InAppNotification {
         payload[PayloadKey.bodyColor] = bodyColor as AnyObject
         payload[PayloadKey.type] = type as AnyObject
         payload[PayloadKey.body] = body as AnyObject
+        var pldDisplayTriggers = [[String: AnyObject]]()
```
Comment (zihejia) [inline Mixpanel/InAppNotification.swift]: what is `pld`, any better name to be used?

File: Mixpanel/DisplayTrigger.swift [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        static let selector = "selector"
+    }
+    
+    static let ANY_EVENT = "$any_event"
+    
+    let event: String?
+    let selector: [String: Any]?
+    let selectorEvaluator: SelectorEvaluator?
+    
+    init?(JSONObject: [String: Any]?) {
```
Comment (zihejia) [inline Mixpanel/DisplayTrigger.swift]: nit: jsonObject,

File: Mixpanel/InAppNotification.swift [hunk 4]
```diff
@@ -55,7 +57,23 @@ class InAppNotification {
         }
 
         guard let extras = object[PayloadKey.extras] as? [String: Any] else {
-            Logger.error(message: "invalid notification extra section")
+            Logger.error(message: "invalid notification extra section \(object)")
+            return nil
+        }
+        
+        if let rawDisplayTriggers = object[PayloadKey.displayTriggers] as? [[String: Any]]? {
+            if let rawDisplayTriggers = rawDisplayTriggers {
```
Comment (zihejia) [inline Mixpanel/InAppNotification.swift]: is it redundant here for assigning `rawDisplayTriggers` again?

File: Mixpanel/SelectorEvaluator.swift [hunk 5]
```diff
@@ -0,0 +1,604 @@
...
+    static let LESS_THAN_OPERATOR = "<"
+    static let LESS_THAN_EQUAL_OPERATOR = "<="
+    // Unary operators
+    static let NOT_OPERATOR = "not"
+    static let DEFINED_OPERATOR = "defined"
+    static let NOT_DEFINED_OPERATOR = "not defined"
+    // Special words
+    static let NOW_LITERAL = "now"
+    
+    let selector: [String:Any]?
```
Comment (zihejia) [inline Mixpanel/SelectorEvaluator.swift]:
```suggestion
    let selector: [String: Any]?
```

File: Mixpanel/SelectorEvaluator.swift [hunk 6]
```diff
@@ -0,0 +1,604 @@
...
+    
+    init?(JSONObject: [String: Any]?) {
```
Comment (zihejia) [inline Mixpanel/SelectorEvaluator.swift]: nit: 
```suggestion
    init?(jsonObject: [String: Any]?) {
```
The same as above

File: Mixpanel/SelectorEvaluator.swift [hunk 7]
```diff
@@ -0,0 +1,604 @@
...
+        guard let op = node[OPERATOR_KEY] as? String, op == NUMBER_OPERATOR else {
+            Logger.error(message: "invalid operator: number")
+            return nil
+        }
+        guard let children = node[CHILDREN_KEY] as? [[String: Any]], children.count == 1 else {
+            Logger.error(message: "invalid operator: number")
+            return nil
+        }
+        
+        return toNumber(value: evaluateNode(node: children[0], properties: properties))
```
Comment (zihejia) [inline Mixpanel/SelectorEvaluator.swift]: would be better to define const for `0` and `1` and avoid using magic number in this case

File: Mixpanel/SelectorEvaluator.swift [hunk 8]
```diff
@@ -0,0 +1,604 @@
...
+        }
+        
+        guard let unit = window[UNIT_KEY] as? String else {
+            Logger.error(message: "missing unit")
+            return nil
+        }
+        let date = getCurrentDate()
+        switch unit {
+        case HOUR_KEY:
+            return date.addingTimeInterval(value.doubleValue*60*60)
```
Comment (zihejia) [inline Mixpanel/SelectorEvaluator.swift]: nit: value.doubleValue * 60 * 60

File: Mixpanel/SelectorEvaluator.swift (hunk 6 above)
Comment (zihejia) [reply Mixpanel/SelectorEvaluator.swift]: can SelectorEvaluator even be a stateless class so you don't even need `init`?

File: Mixpanel/InAppNotification.swift (hunk 4 above)
Comment (madster26) [reply Mixpanel/InAppNotification.swift]: rawDisplayTriggers is optional let me rename it to make it obvious

**Pull request merged**